src/helpers.lisp

```diff
@@ -381,3 +381,26 @@ expected to have the following format:
        (,c-fn (c-struct retval)
               ,@(expand-c-fun-args args))
        retval)))
+
+(defmacro define-print-object (type methods)
+  "Defines PRINT-OBJECT for TYPE. METHODS should contain function names which
+will be called with the object as the parameter. The name of the method and
+it's result will be printed in the output as `:METHOD (method obj)`. The first
+call to PRINT-OBJECT will use PRINT-UNREADABLE-OBJECT, while next-methods will
+only print to the same object.
+
+E.g. (define-print-object sometype (x y)) will print something like `#<SOMETYPE :x 10 :y 20>`"
+  `(defmethod print-object ((obj ,type) out)
+     (let* ((class (find-class ',type))
+            (top (car (closer-mop:compute-class-precedence-list (find-class (type-of obj)))))
```

Comment:
Is there a reason you didn't just reuse `class` here? Is there a situation where `(find-class (type-of obj))` would be different from `(find-class ',type)`?

Reply:
On a similar note, I think if you're not doing `(type-of obj)` here then there are opportunities to push more of the macro code into compile time instead of run time. Not a big issue since this is just a debug function. But the MOP can be slow at runtime so we'd want to be careful with it in other cases.

Reply:
> Is there a reason you didn't just reuse class here? Is there a situation where (find-class (type-of obj)) would be different from (find-class ',type)?

Yes, `obj` will always be the actual type, while `type` will refer to the specialization. So all `call-next-method` calls will be different.

```lisp
(defclass p () ())
(defclass c (p) ())

;; Here the type will always be P, but obj might be C
(defmethod print-object ((obj p) out)
  (format out "~&~A ~A" (find-class (type-of obj)) (find-class 'p))
  (call-next-method))

(defmethod print-object ((obj c) out)
  (format out "~&~A ~A" (find-class (type-of obj)) (find-class 'c))
  ;; Will call print-object where OBJ is C and type is P
  (call-next-method))

(setf some-p (make-instance 'p))
(setf some-c (make-instance 'c))

;; Same
(print-object some-p *standard-output*)
;; Yes, they are the same
;; #<STANDARD-CLASS COMMON-LISP-USER::P> #<STANDARD-CLASS COMMON-LISP-USER::P>#<P {1009E78FE3}>

(print-object some-c *standard-output*)
;; They are the same for print-object for the same class
;; #<STANDARD-CLASS COMMON-LISP-USER::C> #<STANDARD-CLASS COMMON-LISP-USER::C>
;; But when we call the superclass method, they are different
;; #<STANDARD-CLASS COMMON-LISP-USER::C> #<STANDARD-CLASS COMMON-LISP-USER::P>#<C {1009ED8753}>
```

Reply:
> On a similar note, I think if you're not doing (type-of obj) here then there are opportunities to push more of the macro code into compile time instead of run time.

I thought about moving more code out of the macro as I've read many places that it's better to have the macros short and forwarding to functions as early as possible. So I thought about something like the following (which is maybe in the opposite direction..?)

```lisp
(flet ((print-methods (obj class methods out)
         (let ((top (car (closer-mop:compute-class-precedence-list (find-class (type-of obj))))))
           (dolist (method methods)
             (let ((value (handler-case (funcall (symbol-function method) obj)
                                                  (unbound-slot () "#<unbound-slot>"))))
                                     (format out ":~A ~A " method value))))))
  (defmacro define-print-object (type methods)
  "Defines PRINT-OBJECT for TYPE. METHODS should contain function names which
will be called with the object as the parameter. The name of the method and
it's result will be printed in the output as `:METHOD (method obj)`. The first
call to PRINT-OBJECT will use PRINT-UNREADABLE-OBJECT, while next-methods will
only print to the same object.

E.g. (define-print-object sometype (x y)) will print something like `#<SOMETYPE :x 10 :y 20>`"
  `(defmethod print-object ((obj ,type) out)
     (if (eq top class)
         (print-unreadable-object (obj out :type t :identity t)
           (print-methods obj (find-class ',type) ',methods out)
           (call-next-method))
         (progn
           (print-methods obj (find-class ',type) ',methods out)
           (call-next-method))))))
```